sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,234 @@
+     kRomPatchRegionHeaderSizeBytes -                                   \
+     kRomPatchRegionMatchTableSizeBytes -                               \
+     kRomPatchRegionSignatureSizeBytes)
+
+static inline bool rom_patch_lock_valid(uint32_t patch_header) {
+  return bitfield_field32_read(patch_header, kRomPatchLockValid) ==
+         kMultiBitBool4True;
+}
+
+static inline uint8_t rom_patch_revision(uint32_t patch_header) {
+  return (uint8_t)bitfield_field32_read(patch_header, kRomPatchRevision);
+}
+
+static inline uint16_t rom_patch_size_words(uint32_t patch_header) {
+  return (uint16_t)bitfield_field32_read(patch_header, kRomPatchSize);
+}
+
+static inline size_t rom_patch_code_size_words(const rom_patch_t *patch) {
+  HARDENED_CHECK_NE(patch, NULL);
+
+  return ((bitfield_field32_read(patch->header, kRomPatchSize) << 2) -
+          kRomPatchRegionHeaderSizeBytes - kRomPatchRegionMatchTableSizeBytes -
+          kRomPatchRegionSignatureSizeBytes) >>
+         2;
+}
+
+static inline size_t rom_patch_region_enabled(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->m_base != 0 && match->r_base != 0;
+}
+
+static inline bool rom_patch_region_locked(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchLocked) == 1;
+}
+
+static inline size_t rom_patch_region_size(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchPSize);
+}
+
+static inline uint32_t rom_patch_region_r_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->r_base;
+}
+
+static inline uint32_t rom_patch_region_m_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return bitfield_field32_read(match->m_base, kRomPatchMatchMBase);
+}
+
+bool rom_patch_valid(const uintptr_t patch_addr) {
+  if (patch_addr == kRomPatchInvalidAddr || patch_addr < kRomPatchBaseAddr ||
+      patch_addr > kRomPatchMaxAddr) {
+    return false;
+  }
+
+  return rom_patch_size_words(otp_read32(patch_addr)) != 0;
+}
+
+OT_WARN_UNUSED_RESULT uintptr_t rom_patch_latest(void) {
+  size_t next_patch_addr = kRomPatchBaseAddr;
+  size_t latest_patch_addr = kRomPatchInvalidAddr;
+  size_t current_patch_size_words = 0;
+  uint8_t current_patch_revision = 0;
+  uint8_t latest_patch_revision = 0;
+  uint32_t patch_header;
+
+  do {
+    patch_header = otp_read32(next_patch_addr);
+    current_patch_size_words = rom_patch_size_words(patch_header);
+    current_patch_revision = rom_patch_revision(patch_header);
+
+    /*
+     * Check if the current patch is better than the latest, i.e.:
+     * - It is valid, i.e. it's been fully programmed.
+     * - It has a strictly higher revision number.
+     */
+    if (rom_patch_lock_valid(patch_header) &&
```

Comment:
https://github.com/lowRISC/opentitan/blob/integrated_dev/hw/ip/rv_core_ibex/doc/boot-rom-patching.md?plain=1#L369 says that the latest region is the one with highest revision that has the `Program Start` MuBi set. This does not seem to match what the code is doing.

Reply:
The documentation is also not clear on what would happen if there were several patches with the same revision (a case which surely should not happen but should be explicitely be documented and handled by the code I believe)?

Reply:
I see this as a policy decision; if we have an OTP programming failure on one specific device, then the current mechanism may allow us to recover by programming the same patch (=same revision) again at a higher offset.

However, if the ROM code applied all patches that have the highest revision number, then we could incrementally add new patches without having to consume more OTP space by writing out all of the earlier modifications again in the new patch description.